### PR TITLE
A fix for virus scanners preventing directory moves on Windows

### PR DESCRIPTION
**Pull request by ivmarkov**
A fix for 
[this](https://stackoverflow.com/questions/49442860/how-to-solve-the-unable-to-update-binding-error-in-sonar-lint?rq=1) problem.

TL;DR: Some virus scanners have the nasty habit of listening the Windows filesystem for changes and opening a file for scanning immediately after the file had been created by some program (in this case, the SonarLint Eclipse plugin). 

Unfortunately this means that a move or deletion operation on the file (or directory where the file resides) might fail. We've identified this problem in our own code several years ago, and the only workaround is to try to repeat the operation a couple of times until it eventually succeeds.

Fortunately, the code changes for SonarLint are seemingly minimalistic - a very unobtrusive change in FileUtils.java is all that was necessary.

### CODE REVIEW

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java [hunk 1]
```diff
@@ -152,4 +162,47 @@ public static void replaceDir(Consumer<Path> dirContentUpdater, Path target, Pat
...
+   * On Windows, retries the provided IO operation a number of times, in an effort to make the operation succeed.
+   * 
+   * Operations that might fail on Windows are file & directory move, as well as file deletion. These failures 
+   * are typically caused by the virus scanner and/or the Windows Indexing Service. These services tend to open a file handle 
+   * on newly created files in an effort to scan their content.
+   * 
+   * @param runnable the runnable whose execution should be retried
+   */
+  private static void retry(IORunnable runnable) throws IOException {
+    for (int retries = isWindows()? 30: 1; retries > 0; retries--) {
```
Comment (dbmeneses) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: isn't 30 a little excessive? It would take 3seconds to fail, which is a long time in SonarLint environment.

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java [hunk 2]
```diff
@@ -33,9 +34,18 @@
 import java.util.regex.Pattern;
 
 public class FileUtils {
-
+  /**
+   * A simple representation of an IO operation.
+   * An internal interface necessary for the implementation of {@link #retry()}.
+   */
+  private interface IORunnable {
```
Comment (dbmeneses) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: You could mark it with `FunctionalInterface`

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java [hunk 3]
```diff
@@ -152,4 +162,47 @@ public static void replaceDir(Consumer<Path> dirContentUpdater, Path target, Pat
...
+  }
+  
+  /**
+   * A simple check whether the underlying operating system is Windows. 
+   * @return true if the OS is Windows
+   */
+  private static boolean isWindows() {
+	  synchronized (FileUtils.class) {
+		if (windows == null)
+		  windows = System.getProperty("os.name").startsWith("Windows");
```
Comment (dbmeneses) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: This will result in a NPE if for some reason os.name is not defined.

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java [hunk 4]
```diff
@@ -152,4 +162,47 @@ public static void replaceDir(Consumer<Path> dirContentUpdater, Path target, Pat
...
+      }
+    }
+  }
+  
+  /**
+   * A simple check whether the underlying operating system is Windows. 
+   * @return true if the OS is Windows
+   */
+  private static boolean isWindows() {
+	  synchronized (FileUtils.class) {
```
Comment (dbmeneses) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: The value could be simply statically assigned to `windows`. There is no need to calculate it lazily since we know it will always be needed and it avoids to do the synchronization every time.

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 1 above)
Comment (janos-ss) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: I think it would be good to extract `isWindows()? 30: 1` to a `getMaxRetries()` static helper.

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java [hunk 5]
```diff
@@ -152,4 +162,47 @@ public static void replaceDir(Consumer<Path> dirContentUpdater, Path target, Pat
...
+   * 
+   * Operations that might fail on Windows are file & directory move, as well as file deletion. These failures 
+   * are typically caused by the virus scanner and/or the Windows Indexing Service. These services tend to open a file handle 
+   * on newly created files in an effort to scan their content.
+   * 
+   * @param runnable the runnable whose execution should be retried
+   */
+  private static void retry(IORunnable runnable) throws IOException {
+    for (int retries = isWindows()? 30: 1; retries > 0; retries--) {
+      if (retries > 1) {
```
Comment (janos-ss) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: Instead of checking the current retry count in every iteration of the loop, I would use a flag, set after successful run, and then after the loop, run one last time if the flag was not set.

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java [hunk 6]
```diff
@@ -152,4 +162,47 @@ public static void replaceDir(Consumer<Path> dirContentUpdater, Path target, Pat
...
+	      runnable.run();
+	      break;
+	    } catch(AccessDeniedException e) {
+	      // Sleep a bit to give a chance to the virus scanner / Windows Indexing Service to release the opened file handle
+	      try {
+	        Thread.sleep(100);
+	      } catch (InterruptedException ie) {
+	    	// Nothing meaningful to be done here
+	      }
+	    }
```
Comment (janos-ss) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: Please use indentation consistent with the rest of the file. (2 spaces per level)

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 1 above)
Comment (ivmarkov) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: We don't expect it to fail. *We expect it so succeed*. Moreover, it would succeed on the second or third retry at most which is *100ms wait time or 200ms at most*. Always, or almost always. But if - for whatever reason - once in a lifetime - the PC is under a heavy load, and McAfee is stressed to complete its virus checking on time, we don't want to report failure. The user would then wait for 3 seconds. Not a big deal IMHO.

I think if you want to continue arguing over the same lines, you should concentrate on the Thread.sleep timeout, rather than on the number of retry, as at least one Thread.sleep call will always happen on Windows, if McAfee is installed and screwing things for us.

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 4 above)
Comment (ivmarkov) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: Obvious in retrospective. Will fix it accordingly, thanks!

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 5 above)
Comment (ivmarkov) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: That was how the code looked like originally. A question of preference, if you ask me. But since I'm anyway fixing stuff, I can fix that as well. Anything against early exits/returns?

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 6 above)
Comment (ivmarkov) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: Sorry for that - missed that one (we still use the old 4 spaces per level in our own codebase).

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 3 above)
Comment (dbmeneses) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: I agree

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java (hunk 1 above)
Comment (dbmeneses) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: Alright then.
Just a note: here you are assuming that when an `AccessDeniedException` happens, it's a temporary situation. However, I imagine this happening due to an actual lack of permissions and in this case we want to fail as fast as possible, hence my comment. In this situation what matters is the total timeout, not the number retries or the duration of each sleep.

Comment (ivmarkov) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java]: Right... yet: what is the chance the user not to have permissions in - say - the temporary space where these files reside.

Moreover, if there is a general permissions' issue, I don't think you can recover from that. In that case, most likely the stack will be unrolled and a general error (maybe even internal error?) will appear in Eclipse, VS etc. The fact that it will appear 3 seconds later is probably not the end of the world.

**Comment by ivmarkov** [comment]
Will this ever get merged into the SonarLint plugin for Eclipse? 
More than two months had passed, and the latest Eclipse plugin is still failing on file move operations.

**Comment by ivmarkov** [comment]
Hello? 
Sorry for being a bit direct, but - do you guys seriously believe, that your Eclipse plugin will ever get used in a corporate environment with this issue not fixed?

**Comment by janos-ss** [comment]
Hi @ivmarkov, sorry for not replying to your previous message.

This PR has been merged a while ago, but it just missed the previous release of SonarLint for Eclipse. It will be part of the next one, which should be coming soon, within a month or so.
